### PR TITLE
[CDAP-19221] Conversion of control center tests from cypress to selenium

### PR DESCRIPTION
**Pull request by radhikav1**
# [CDAP-19221] Convert controlcenter.spec.ts tests from cypress to selenium

## Description
Conversion of controlcenter tests to selenium and added required helper methods.

## PR Type
- [ ] Bug Fix
- [ ] Feature
- [ ] Build Fix
- [x] Testing
- [ ] General Improvement
- [ ] Cherry Pick

## Links
Jira: [CDAP-19221](https://cdap.atlassian.net/browse/CDAP-19221)

## Test Plan
Included in  cdap ui e2e tests.

## Screenshots
n/a

[CDAP-19221]: https://cdap.atlassian.net/browse/CDAP-19221?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/e2e-test/java/io/cdap/cdap/ui/utils/Constants.java [hunk 1]
```diff
@@ -32,6 +32,9 @@ private static double getRandomArbitrary(int min, int max) {
   public static final String CONFIGURATION_URL = "http://localhost:11011/cdap/administration/configuration";
   public static final String SYSTEM_PROFILES_CREATE_URL = "http://localhost:11011/cdap/ns/system/profiles/create";
   public static final String DEFAULT_PROFILES_CREATE_URL = "http://localhost:11011/cdap/ns/default/profiles/create";
+  public static final String CONTROL_CENTER_URL = "http://localhost:11011/cdap/ns/default/control";
```
Comment (mrahanjam) [inline src/e2e-test/java/io/cdap/cdap/ui/utils/Constants.java]: You don't need to add a constant for every URL. you can create a base studio URL like `http://localhost:11011/cdap/ns/default/` and use `BASE_STUDIO_URL + "profiles/create"`

File: src/e2e-test/java/io/cdap/cdap/ui/utils/Helper.java [hunk 2]
```diff
@@ -87,4 +102,74 @@ public static String getNodeSelectorFromNodeIndentifier (NodeInfo node) {
...
+    ElementHelper.clickOnElement(locateElementById("resource-center-btn"));
+    ElementHelper.clickOnElement(locateElementById("create-pipeline-link"));
+    if(!SeleniumDriver.getDriver().getCurrentUrl().contains("studio")) {
+      throw new RuntimeException("URL redirection for pipeline creation unsuccessful.");
+    }
+    WaitHelper.waitForPageToLoad();
+    WebElement upload_file = SeleniumDriver.getDriver()
+            .findElement(By.xpath("//*[@id='pipeline-import-config-link']/input[@type='file']"));
+
+    File pipelineJSONFile = new File("src/e2e-test/fixtures/" + filename);
```
Comment (mrahanjam) [inline src/e2e-test/java/io/cdap/cdap/ui/utils/Helper.java]: you can create a constant for `fixtures dir` in `Constants` class like `FIXTURES_DIR="src/e2e-test/fixtures/"`

File: src/e2e-test/java/io/cdap/cdap/ui/utils/Helper.java [hunk 3]
```diff
@@ -87,4 +102,74 @@ public static String getNodeSelectorFromNodeIndentifier (NodeInfo node) {
...
+    pipeline_name_input.clear();
+    pipeline_name_input.sendKeys(pipelineName);
+    pipeline_name_input.sendKeys(Keys.RETURN);
+
+//    WebElement deployPipelineElm =  WaitHelper.waitForElementToBeClickable(
+//                    locateElementByCssSelector(getCssSelectorByDataTestId("deploy-pipeline")), 1000);
+//    deployPipelineElm.click();
+// Below is the workaround as above(commented block) is not working
+    WebElement deployPipelineElm = locateElementByCssSelector(getCssSelectorByDataTestId("deploy-pipeline"));
+    JavascriptExecutor jse = (JavascriptExecutor) SeleniumDriver.getDriver();
```
Comment (mrahanjam) [inline src/e2e-test/java/io/cdap/cdap/ui/utils/Helper.java]: Why do you need a Javascript executor for a click action?

File: src/e2e-test/java/io/cdap/cdap/ui/utils/Helper.java [hunk 4]
```diff
@@ -87,4 +102,74 @@ public static String getNodeSelectorFromNodeIndentifier (NodeInfo node) {
...
+
+//    ElementHelper.clickOnElement(
+//            WaitHelper.waitForElementToBeClickable(locateElementByCssSelector("div[class*='pipeline-name']"))
+//    );
+    // Below is the workaround as above(commented block) is not working
+    try {
+      WebElement pipelineNameContainer = locateElementByCssSelector("div[class*='pipeline-name']");
+      ElementHelper.clickOnElement(pipelineNameContainer);
+    } catch(org.openqa.selenium.StaleElementReferenceException ex) {
+      WebElement pipelineNameContainer = locateElementByCssSelector("div[class*='pipeline-name']");
```
Comment (mrahanjam) [inline src/e2e-test/java/io/cdap/cdap/ui/utils/Helper.java]: This also feels weird as you are the same steps in the try block again in the catch block (what if it fails again?)

File: src/e2e-test/java/io/cdap/cdap/ui/utils/Constants.java (hunk 1 above)
Comment (radhikav1) [reply src/e2e-test/java/io/cdap/cdap/ui/utils/Constants.java]: Addressed

File: src/e2e-test/java/io/cdap/cdap/ui/utils/Helper.java (hunk 2 above)
Comment (radhikav1) [reply src/e2e-test/java/io/cdap/cdap/ui/utils/Helper.java]: yes, done

File: src/e2e-test/java/io/cdap/cdap/ui/utils/Helper.java (hunk 4 above)
Comment (radhikav1) [reply src/e2e-test/java/io/cdap/cdap/ui/utils/Helper.java]: Found alternative and fixed it

File: src/e2e-test/java/io/cdap/cdap/ui/utils/Helper.java (hunk 3 above)
Comment (radhikav1) [reply src/e2e-test/java/io/cdap/cdap/ui/utils/Helper.java]: Changed with some helper functions and JS executor is not needed now.

File: src/e2e-test/features/controlcenter.feature [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+Feature: Control Center - filter and find
+
+  @CONTROL_CENTER_FILTER_FIND
+  Scenario: Using control center to see and filter things
+    When Deploy and test pipeline "test_pipeline2_fll_airport" with pipeline JSON file "fll_airport_pipeline2.json"
+    When Open control center page
+    Then Find the pipeline "test_pipeline2_fll_airport" in the results
+    Then Find filter dropdown and unselect "applications"
+    Then Verify that pipeline "test_pipeline2_fll_airport" should not present
+    Then Clean up pipeline "test_pipeline2_fll_airport" which is created for testing
```
Comment (mrahanjam) [inline src/e2e-test/features/controlcenter.feature]: nit: Missing blank line at EOF

**Review by mrahanjam — APPROVED** [review]
Left a nit, overall LGTM. Great work 🚀

**Comment by mrahanjam** [comment]
There's still an issue with the `cdap-build` github action (used for building the sandbox which our github action uses). That's why your PR s=is failing the build step. Once that's resolved I'll re-trigger the build here so you can merge afterwards.

File: src/e2e-test/features/controlcenter.feature (hunk 5 above)
Comment (radhikav1) [reply src/e2e-test/features/controlcenter.feature]: Done

File: src/e2e-test/java/io/cdap/cdap/ui/stepsdesign/ControlCenter.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
+package io.cdap.cdap.ui.stepsdesign;
```
Comment (sumengwang) [inline src/e2e-test/java/io/cdap/cdap/ui/stepsdesign/ControlCenter.java:17]: copyrights header

File: app/cdap/components/EntityCard/index.js [hunk 7]
```diff
@@ -137,6 +137,7 @@ export default class EntityCard extends Component {
               className={classnames({ 'with-version': this.props.entity.version })}
               title={this.props.entity.id}
               data-cy={`${this.props.entity.id}-header`}
+              data-testid={`${this.props.entity.id}-header`}
```
Comment (seanfeld) [inline app/cdap/components/EntityCard/index.js:140]: @mrahanjam we're switching over to using data-testid? if so why do we still need data-cy?

File: src/e2e-test/java/io/cdap/cdap/ui/stepsdesign/ControlCenter.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+        ElementHelper.selectCheckbox(Helper.locateElementByCssSelector(
+                Helper.getCssSelectorByDataTestId("Applications-input")));
+    }
+
+    @Then("Verify that pipeline {string} should not present")
+    public void verifyPipeLineNotDisplayed(String pipelineName) {
+        String cssSelector = "div[class*='entities-all-list-container'] > "
+                + Helper.getCssSelectorByDataTestId(pipelineName + "-header");
+        boolean isPipelineFound = Helper.isElementExists(cssSelector);
+        if(isPipelineFound) {
```
Comment (seanfeld) [inline src/e2e-test/java/io/cdap/cdap/ui/stepsdesign/ControlCenter.java]: space after if

File: app/cdap/components/EntityCard/index.js (hunk 7 above)
Comment (mrahanjam) [reply app/cdap/components/EntityCard/index.js:140]: Cypress tests are still being used in cloud build, once we move away from using cypress completely, we can remove data-cy tags.

File: src/e2e-test/java/io/cdap/cdap/ui/utils/Helper.java [hunk 9]
```diff
@@ -87,4 +103,61 @@ public static String getNodeSelectorFromNodeIndentifier (NodeInfo node) {
       node.getNodeType() + "-" +
       node.getNodeId() + "\"]";
   }
+
+  public static void deployAndTestPipeline(String filename, String pipelineName) {
+    SeleniumDriver.openPage(Constants.BASE_STUDIO_URL + "pipelines");
+    WaitHelper.waitForPageToLoad();
+    ElementHelper.clickOnElement(locateElementById("resource-center-btn"));
+    ElementHelper.clickOnElement(locateElementById("create-pipeline-link"));
+    if(!SeleniumDriver.getDriver().getCurrentUrl().contains("studio")) {
```
Comment (seanfeld) [inline src/e2e-test/java/io/cdap/cdap/ui/utils/Helper.java]: i couldn't find the exact rule but for java inside g3 but it looks like the preferred way is having a space after if statements https://google.github.io/styleguide/javaguide.html#s4.1.2-blocks-k-r-style

**Review by sumengwang — APPROVED** [review]

File: src/e2e-test/java/io/cdap/cdap/ui/utils/Helper.java (hunk 9 above)
Comment (radhikav1) [reply src/e2e-test/java/io/cdap/cdap/ui/utils/Helper.java]: yeah, changes done as suggested

File: src/e2e-test/java/io/cdap/cdap/ui/stepsdesign/ControlCenter.java (hunk 8 above)
Comment (radhikav1) [reply src/e2e-test/java/io/cdap/cdap/ui/stepsdesign/ControlCenter.java]: done

**Pull request merged**